doc/specs/#190 - Proxy Support.md

```diff
@@ -0,0 +1,98 @@
+---
+author: Flor Chacon @florelis
+created on: 2024-02-07
+last updated: 2024-02-07
+issue id: 190
+---
+
+# Proxy support
+
+For [#190](https://github.com/microsoft/winget-cli/issues/190)
+
+## Abstract
+
+This spec describes a feature to specify a proxy for winget to use when connecting to the internet.
+
+## Solution Design
+
+A new command line argument will be added to specify a proxy to use during a particular invocation of winget.
+This functionality will first need to be enabled through an admin setting, similar to local manifests or hash override.
+
+An option to set a default proxy to use on every flow will be added, but it will require administrator permissions to be set.
+
+New Group Policy will also be added for IT admins to control the use of proxies.
+The policies will be similar to those we already have for sources, so that a specific proxy can be required or only a predefined set of proxies can be allowed.
+
+Proxies will not be used for the configuration features for now.
+Proxy settings in winget will not affect the behavior of installers themselves, so an installation may still generate traffic outside of the proxy.
+
+The proxy will be used to download the installers and access the sources.
+
+Since the APIs used for Delivery Optimization and MSIX deployment do not provide a way to specify a custom proxy, if a proxy is specified, we will change the use of those APIs to accomodate the proxy.
+For Delivery Optimization, a proxy will force use of WinINet instead.
+For MSIX deployment, the packages will be fully downloaded before deploying; this will require more network traffic than a streaming install of only the required bits.
+
+## UI/UX Design
+
+We will add a command line argument taking the URI to the proxy.
+A separate argument will be available to disable the use of proxy if there is a default set.
+Both of these arguments will be disabled by default and require admin privileges to enable.
+
+```
+> winget settings --enable ProxyCommandLineArgument
+> winget install Contoso.App --proxy https://127.0.0.1:2345
+> winget install Contoso.App --no-proxy
+```
+
+To configure the default proxy, a new `proxy` subcommand will be added to the `settings` command, with options to `set` and `reset` the default.
+This will require admin privileges and does not require `ProxyCommandLineArgument` to be enabled.
+
+```
+> winget settings proxy set https://127.0.0.1:2345
+> winget settings proxy reset
+```
```

Comment:
The way I actually implemented this in #4203 was having `set` and `reset` subcommands for `settings` so that we can do
`winget settings set [--setting] <SomeAdminSetting> [--value] <SomeValue>` or `winget settings set [--setting] <SomeAdminSetting>`. Currently the only admin setting that can be used there would be `DefaultProxy`, but I think it could be extensible.